docs/architecture/adr-067-mempool-refactor.md

```diff
@@ -0,0 +1,137 @@
+prioritize transactions in a flexible and performant manner.
+
+## Alternative Approaches
+
+When considering which approach to take for a priority-based flexible and
+performant mempool, there are two core candidates. The first candidate in less
+invasive in the required  set of protocol and implementation changes, which
+simply extends the existing `CheckTx` ABCI method. The second candidate essentially
+involves the introduction of new ABCI method(s) and would require a higher degree
+of complexity in protocol and implementation changes, some of which may either
+overlap or conflict with the upcoming introduction of [ABCI++](https://github.com/tendermint/spec/blob/master/rfc/004-abci%2B%2B.md).
+
+For more information on the various approaches and proposals, please see the
+[mempool discussion](https://github.com/tendermint/tendermint/discussions/6295).
+
+## Decision
+
+To incorporate a priority-based flexible and performant mempool in Tendermint Core,
+we will introduce new fields, `priority`, `sender`, and `nonce` , into the
+`ResponseCheckTx` type and augment the existing mempool data structure to
+facilitate prioritization of uncommitted transactions in addition to extended
+functionality such as replace-by-priority and allowing multiple transactions to
+exist from the same sender with varying priorities.
+
+## Detailed Design
+
+### CheckTx
+
+We introduce the following new fields into the `ResponseCheckTx` type:
+
+```diff
+message ResponseCheckTx {
+  uint32         code       = 1;
+  bytes          data       = 2;
+  string         log        = 3;  // nondeterministic
+  string         info       = 4;  // nondeterministic
+  int64          gas_wanted = 5 [json_name = "gas_wanted"];
+  int64          gas_used   = 6 [json_name = "gas_used"];
+  repeated Event events     = 7 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "events,omitempty"];
+  string         codespace  = 8;
++ int64          priority   = 9;
++ string         sender     = 10;
++ int64          nonce      = 11;
+}
+```
+
+It is entirely up the application in determining how these fields are populated
+and with what values, e.g. the `sender` could be the signer  and fee payer 
+the transaction, the `priority` could be the cumulative sum of the fee(s), and
+the `nonce` could be the signer's sequence number/nonce.
+
+Only `sender` is required, while `priority` and `nonce` can be omitted which
+would result in using the default value of zero.
+
+### Mempool
+
+The existing concurrent-safe linked-list will be removed entirely in favor of a
+thread-safe map of `<sender:[]Tx>`, i.e a mapping from `sender` to a list of `Tx`
+objects, where each `Tx` is ordered by monotonically increasing, but not
+necessarily consecutive, `nonce` values. For example, the sender `Alice`, whose
+current committed nonce value is `3`, could have transactions in this map with
+nonce values `<4,5,6,8>`, where only the transaction with nonce `4` could be
```

Comment:
Does the concept of a tx nonce already exist in Tendermint and/or the SDK, or is this a new thing?

Reply:
It exists in the SDK and is referred to as sequence number. While not ideal to introduce to Tendermint as it kind of infers an account-based model, I still think it's flexible enough because applications can just choose to ignore this or map nonce onto something else.

Reply:
Nonce is being removed from this ADR entirely 👍